### PR TITLE
Raise exception for negative order derivatives

### LINKED ISSUE #13873: Derivatives of negative order should be rejected
This seems wrong:
```
>>> sin(x).diff(x, -1)
sin(x)
```
It's meaningless and it doesn't even print correctly:
```
>>> pprint(Derivative(sin(x), (x, -1)))
d         
──(sin(x))
dx 
``` 

An attempt to create a Derivative with negative order of differentiation should raise a ValueError in the `__new__` method of `Derivative` class.  I imagine the check being added [here](https://github.com/sympy/sympy/blob/master/sympy/core/function.py#L1176):

```
if count.is_negative:
    raise ValueError("Negative order derivatives are not allowed")
```

This is not entirely busywork because a user may have determined derivative order programmatically, and had a bug somewhere. It's better to inform them of the error than to ignore it.

### PR DESCRIPTION
**Pull request by himanshuladia**
Fixes #13873 

Currently, diff(sin(x), (x, -1)) returns sin(x) which is basically meaningless. An if statement is added to check for negative order derivatives and raise a ValueError. Corresponding tests are also added.

### CODE REVIEW

**Comment by asmeurer** [comment]
A better check would be `count.is_positive is False`, as complex counts have the same problem.

**Comment by normalhuman** [comment]
Good point, but zero order is okay, so I suggest `count_is_nonnegative`:
```
if not all(count.is_nonnegative for v, count in variable_count): 
```

**Comment by himanshuladia** [comment]
I replaced 
`if any(count.is_negative for v,count in variable_count)` 
with
`if not all(count.is_nonnegative for v, count in variable_count): `
A lot of tests which defined `n = symbols('n', integer=True)` are failing now. I guess they require to add `nonnegative=True` parameter. Shall I go for it?

**Comment by normalhuman** [comment]
Yes, I think the tests should be corrected in this way. It is strange that n is declared integer but not declared nonnegative there. Be sure to check bin/doctest too, there may be some there.

Also, instead of `ValueError("Derivative order invalid")` I would say more clearly what the issue is: `ValueError("Derivative order must be a nonnegative integer")`

**Comment by himanshuladia** [comment]
I got rid of all of them but one.

`sympy/utilities/tests/test_pickling.py:test_core_undefinedfunctions ______
  File "/home/himanshuladia/Desktop/sympy/sympy/utilities/tests/test_pickling.py", line 154, in test_core_undefinedfunctions
    check(f, exclude=exclude)
  File "/home/himanshuladia/Desktop/sympy/sympy/utilities/tests/test_pickling.py", line 77, in check
    c(a, b, d1)
  File "/home/himanshuladia/Desktop/sympy/sympy/utilities/tests/test_pickling.py", line 76, in c
    assert getattr(b, i) == attr, "%s != %s" % (getattr(b, i), attr)
AssertionError: {'__doc__': None, '_extra_kwargs': {}, '__module__': None, '_explicit_class_assumptions': {}, 'default_assumptions': {}, '_prop_handler': {'negative': <function _eval_is_negative at 0x7fee1d914d08>, 'commutative': <function _eval_is_commutative at 0x7fee1d9148c8>, 'positive': <function _eval_is_positive at 0x7fee1d914ae8>, 'complex': <function _eval_is_complex at 0x7fee1d914b70>}, '_nargs': None} != {'_extra_kwargs': {}, '__module__': None, '__doc__': None, '_explicit_class_assumptions': {}, 'default_assumptions': {}, '_prop_handler': {'negative': <function Expr._eval_is_negative at 0x7fee609df510>, 'commutative': <function Function._eval_is_commutative at 0x7fee606b68c8>, 'positive': <function Expr._eval_is_positive at 0x7fee609df488>, 'complex': <function Function._eval_is_complex at 0x7fee606b6950>}, '_nargs': None}`

I just am not able to understand the assertion error.

**Comment by normalhuman** [comment]
The error in pickling is unrelated to your PR, something there is broken and you can ignore it. It shouldn't show up on Travis. More importantly, I think asmeurer is right: we shouldn't insist on users declaring nonnegativity (it's fine if you already changed that in the tests, because it still makes sense to do. So instead of 
```
if not all(count.is_nonnegative for v, count in variable_count):
```
which requires that counts be **known** to be nonnegative, use 
```
if any(count.is_nonnegative is False or count_is_integer is False \
    or count.is_finite is False for v, count in variable_count):
```
which throws an error only if there is some knowledge that a count fails to be a nonnegative integer. 

(Interestingly, the last check is not redundant: intinity (oo) passes the first two as its integer status is None.)

**Comment by Upabjojr** [comment]
Shouldn't derivatives of negative number of times be integrals?

File: sympy/core/tests/test_diff.py [hunk 1]
```diff
@@ -109,7 +109,7 @@ def test_diff_nth_derivative():
     f =  Function("f")
     x = Symbol("x")
     y = Symbol("y")
-    n = Symbol("n", integer=True)
+    n = Symbol("n", integer=True, nonnegative=True)
```
Comment (Upabjojr) [inline sympy/core/tests/test_diff.py]: I wouldn't modify this variable. If the assumption is not specified, I would let the code compute the derivative.

**Comment by Upabjojr** [comment]
What about a warning instead of raising an exception?

**Comment by himanshuladia** [comment]
No, derivatives of negative order shouldn't be evaluated as integrals. As @normalhuman pointed out, that's a hairy subject. You can check this out [here](https://math.stackexchange.com/questions/1503551/derivative-of-a-negative-order/1503626#1503626) And since this isn't right, one should explicitly declare the order as nonnegative. If that'd be negative, in my opinion, it should raise an error. There's no point in showing a warning if that doesn't makes sense.

**Comment by ylemkimon** [comment]
I think if the user provides a negative number, he or she is probably intending to perform an integration. Checking existence and uniqueness and evaluating it should be up to the user. [Mathematica](http://reference.wolfram.com/language/ref/Derivative.html) seems to do this, [Maple](https://www.maplesoft.com/support/help/Maple/view.aspx?path=diff/symbolicorder) does this for symbolic order, not sure it does for an integer literal, [Maxima](http://maxima.sourceforge.net/docs/manual/maxima_18.html) doesn't.

**Comment by himanshuladia** [comment]
That's right. I think if a user explicitly provides a negative order for instance `diff(sin(x), (x, -1))`, it should perform an integration. 

But something like 
`n = symbols('n', integer=True)
diff(sin(x), (x,n))`
should implicitly perform a derivative only, as @Upabjojr said.

**Comment by normalhuman** [comment]
How about this: we don't raise any exceptions or even warnings, but just leave the derivative un-evaluated. My objection in the original issue was to `diff(sin(x), x, -1)` returning `sin(x)`.

**Comment by Upabjojr** [comment]
Unevaluated form works for me.

**Comment by asrdav** [comment]
Is this issue still open?

**Comment by normalhuman** [comment]
@asrdav It seems in the process of reviewing this PR the approach toward the issue changed, so that it is no longer deemed necessary to raise an exception.  

It is still a problem that `sin(x).diff(x, -1)` returns `sin(x)`, though.  It should remain unevaluated as `Derivative(sin(x), (x, -1))` until and unless something is implemented for negative orders. 

It's possible that the API changes for n-th order derivative that @Upabjojr has been working on will change the context of this issue, so I think it's best to postpone addressing the issue for a while.

**Comment by Upabjojr** [comment]
I have fixed this in https://github.com/sympy/sympy/pull/13972

**Pull request closed without merging**